src/new_skills/NewSkill.cpp

```diff
@@ -0,0 +1,135 @@
+            return Status::Failure;
+        }
+        if (robot->id == -1) {
+            std::cerr << "The robot id is" << robot->id << std::endl;
+            return Status::Failure;
+        }
+        if (!ball) {
+            return Status::Waiting;
+        }
+        return onUpdate();
+    }
+
+    void NewSkill::initialize() {
+        // TODO: be able to get robot from dealer
+        // robot = dealer.distributions.at(role).robot;
+        ball = world->getBall();
+        if (!robot || robot->id == -1) return;
+        if (!ball) return;
+        refreshRobotCommand();
+        onInitialize();
+    }
+
+    void NewSkill::terminate(Status s) {
+        if (!init) {
+            return;
+        }
+        init = false;
+        if (!robot || robot->id == -1) return;
+        if (!ball) return;
+        refreshRobotPositionControllers();
+        refreshRobotCommand();
+        onTerminate(s);
+    }
+
+    proto::RobotCommand NewSkill::rotateRobotCommand(proto::RobotCommand &cmd) {
+        proto::RobotCommand output = cmd;
+        output.mutable_vel()->set_x(-cmd.vel().x());
+        output.mutable_vel()->set_y(-cmd.vel().y());
+        output.set_w(static_cast<float>(control::ControlUtils::constrainAngle(cmd.w() + M_PI)));
+        return output;
+    }
+
+    void NewSkill::refreshRobotCommand() {
+        proto::RobotCommand emptyCmd;
+        emptyCmd.set_use_angle(true);
+        emptyCmd.set_id(robot ? robot->id : -1);
+        emptyCmd.set_geneva_state(0);
+        command = emptyCmd;
+    }
+
+/// Velocity and acceleration limiters used on command. TODO: this function is nasty, need to actually make it do what it says it does
+    void NewSkill::limitRobotCommand() {
+        bool isDefendPenaltyState = rtt::ai::GameStateManager::getCurrentGameState().keeperStrategyName == "keeper_penalty_defend_tactic";
+        bool isKeeper = command.id() == robotDealer::RobotDealer::getKeeperID();
+
+        auto limitedVel = Vector2(command.vel().x(), command.vel().y());
+        limitedVel = control::ControlUtils::velocityLimiter(limitedVel);
+        if (!(isDefendPenaltyState && isKeeper)) {
+            limitedVel = control::ControlUtils::accelerationLimiter(limitedVel, robot->getPidPreviousVel(), command.w());
+        }
+        robot->setPidPreviousVel(limitedVel);
+        if (std::isnan(limitedVel.x) || std::isnan(limitedVel.y)) {
+            std::cout << "ERROR in NewSkill::limitRobotCommand() : The robot has a velocity of NaN: " << node_name().c_str() << "!"
+                      << "  robot  " << robot->id << std::endl;
+            robot->setPidPreviousVel(robot->vel);
+        }
+        command.mutable_vel()->set_x(limitedVel.x);
+        command.mutable_vel()->set_y(limitedVel.y);
+    }
+
+    void NewSkill::refreshRobotPositionControllers() {
+        robot->resetNumTreePosControl();
+        robot->resetShotController();
+        robot->resetBallHandlePosControl();
+        robot->resetBasicPosControl();
```

Comment:
I still think that the controllers should not be accessed through the robot object

Reply:
I don't have a view on this, but that is also not the purpose of this PR. I'm trying to get a stub moving for a new implementation of role/skill that is able to work with the new dealer.

Reply:
I agree with Emiel on this

Reply:
I am interested why @emielsteerneman and @timovdk think this. Last year we explicitly moved the controllers from skills towards robot, because otherwise the PID controllers would reset when switching skills, resulting in spectacular collisions and backflip robots. Therefore, from a control viewpoint it makes a lot of sense to put the controllers in Robot. 

--> my main point is that each robot should have it's own set of controllers that stay alive for the whole runtime duration of AI

Reply:
Since with the RobotViews the idea was to have data only, it does not immediately make sense to have controllers in there as well. It is more a structural difference in my opinion.

Reply:
I agree with Timo and Emiel. The controllers need to be long lived but this is not the place to do so. Robots can get deleted or replaced in world whilst this would have no implications for control at all.

 I would much rather see a separate entity that manages the controllers somewhere else than here. Whether this is the right place and time to fix it, I don't know.

Reply:
(just throwing in some thoughts)

@timovdk That makes sense indeed. On the other hand, the controller does contain robot-specific data. In fact, a controller is a set of functions with some robot-specific parameters. maybe the controllers can be made into static functions while the input data for these controllers is stored in robot? That would have the implication that robot is mutable however. I agree it is all not ideal.

Another option is to have a set of controllers for each robot which is stored in (for example) a map or some ControlManager. this object then needs to live in applicationmanager and be passed around. This would already be quite nice. But maybe there is a better solution...

Reply:
I would personally prefer the second option but as @mrlukasbos said, there might even be better options, that is why I am also not sure whether this is the right place and time to fix this.

Reply:
I think I also disagree with putting the controllers in the robot class, I don't think it makes sense to do so.

Not sure what a good solution would be though